pkg/kube/dialer.go

```diff
@@ -0,0 +1,217 @@
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+// podAddr implements net.Addr for kubernetes pod connections
+type podAddr struct {
+	PodName string
+	Port    string
+}
+
+func (podAddr) Network() string {
+	return "tcp"
+}
+
+func (a podAddr) String() string {
+	return fmt.Sprintf("%s:%s", a.PodName, a.Port)
+}
+
+// PodDialer uses kubernetes' portforwarding protocol to create a net.Conn
+// to a pod in the given kubernetes clusters
+type PodDialer struct {
+	Namespace string
+	Config    *rest.Config
+	ClientSet kubernetes.Interface
+	Transport http.RoundTripper
+	Upgrader  spdy.Upgrader
+
+	mu             sync.Mutex
+	requestCounter int
+	dialers        map[string]httpstream.Dialer
+}
+
+// NewPodDialer creates a PodDailer that allows for a database connection to flow
+// through a connection created by a kube-proxy like connection.
+func NewPodDialer(config *rest.Config, namespace string) (*PodDialer, error) {
+	transport, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PodDialer{
+		Config:    config,
+		Namespace: namespace,
+		ClientSet: kubernetes.NewForConfigOrDie(config),
+		Transport: transport,
+		Upgrader:  upgrader,
+		dialers:   make(map[string]httpstream.Dialer),
+	}, nil
+}
+
+func (k *PodDialer) nextRequestID() int {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+	k.requestCounter += 1
+	return k.requestCounter
+}
+
+func (d *PodDialer) dialerForPod(podName string) httpstream.Dialer {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// Reuse any dialers that are already created
+	if dialer, ok := d.dialers[podName]; ok {
+		return dialer
+	}
+
+	podShortName := strings.Split(podName, ".")
+
+	// Build a raw request so we can extract the URL
+	req := d.ClientSet.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Namespace(d.Namespace).
+		Name(podShortName[0]).
+		SubResource("portforward")
+
+	dialer := spdy.NewDialer(d.Upgrader, &http.Client{Transport: d.Transport}, "POST", req.URL())
+
+	d.dialers[podName] = dialer
+
+	return dialer
+}
+
+// DialTimeout implements pq.Dialer
+func (k *PodDialer) DialTimeout(network, addr string, d time.Duration) (net.Conn, error) {
+	return k.Dial(network, addr)
+}
+
+func (k *PodDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	return k.Dial(network, addr)
+}
+
+// Dial connects to a port in a kubernetes pod specified by addr. network must be TCP
+// Implmentation adapted from:
+//	https://github.com/kubernetes/kubernetes/blob/27c70773add99e43464a4e525e3bddfc8b602a3d/staging/src/k8s.io/client-go/tools/portforward/portforward.go
+//	https://github.com/kubernetes/kubernetes/blob/27c70773add99e43464a4e525e3bddfc8b602a3d/staging/src/k8s.io/kubectl/pkg/cmd/portforward/portforward.go
+func (k *PodDialer) Dial(network, addr string) (net.Conn, error) {
```

Comment:
There are a couple changes to the PodDialer, we found that it was leaking goroutines as `streamConn` is never closed. I'll send you the patch!

Reply:
Please let me know